port to Zig 0.15

pluiedev opened the pull request:
Or: "How I learned to stopped worrying and love `std.Io.Writer.Allocating`"

Progress:
- [X] Build system compiles
- [X] libghostty compiles
- [x] Test suite compiles
- [x] Test suite all green
- [x] Benchmarks compile
- [ ] Benchmarks run and perform well against 0.14
- [x] All CLI commands compile
- [x] GTK-NG + OpenGL compiles
- [x] macOS app + Metal compiles
  - Blocked by zig-objc
  - I don't have a Mac ;)
- [x] Packaging manifests/metadata/instructions updated
- [x] CI all green
  - FreeBSD builds disabled for now since Ports has yet to update to 0.15

Closes #8361

pluiedev committed:
nix: update flake

pluiedev committed:
build: port to Zig 0.15

pluiedev committed:
build: `format` considered evil; all build script errors fixed

pluiedev committed:
build: update *most* dependencies to Zig 0.15

Except libvaxis since that thing depends on SO many dependencies it's
absolutely unfeasible to update all of them on my own, so I'm relying on
Tim to do the job for me for now

pluiedev committed:
core: port to Zig 0.15

libghostty can now be built with 0.15!

pluiedev committed:
gtk-ng: port to Zig 0.15

pluiedev committed:
treewide: ArrayListUnmanaged -> ArrayList

pluiedev committed:
treewide: std.io -> std.Io

pluiedev committed:
core: all tests compile, most run successfully

jcollie commented on src/apprt/gtk/App.zig:
```diff
@@ -101,7 +101,7 @@ transient_cgroup_base: ?[]const u8 = null,
 
 /// Providers for loading custom stylesheets defined by user
-custom_css_providers: std.ArrayListUnmanaged(*gtk.CssProvider) = .{},
+custom_css_providers: std.ArrayList(*gtk.CssProvider) = .{},
```
`.empty` for consistency

jcollie commented on src/config/Config.zig:
```diff
@@ -3143,7 +3143,7 @@ _conditional_set: std.EnumSet(conditional.Key) = .{},
 /// without reopening the files. This is used in very specific cases such
 /// as loadTheme which has more details on why.
-_replay_steps: std.ArrayListUnmanaged(Replay.Step) = .{},
+_replay_steps: std.ArrayList(Replay.Step) = .{},
```
Another one. Muahahaha!

jcollie commented:
Upstream on zig-gobject is working on porting to Zig 0.15: https://github.com/ianprime0509/zig-gobject/pull/119

pluiedev committed:
treewide: de-anytype-ify

pluiedev committed:
crash: nuke minidump

Writergate would require rewriting this entire thing.

pluiedev committed:
build: force GhosttyExe to build with LLVM

Otherwise the dynamic libraries aren't linked correctly

pluiedev committed:
core: All Unit Tests Green

Seriously some of these tests never should have worked

pluiedev committed:
zig fmt

Apparently Zig 0.15 now forces an order for `@ptrCast` and ``@constCast`?
Interesting

pluiedev committed:
build/webgen: port to Zig 0.15

pluiedev committed:
build: fix sentry, benchmark, flatpak builds

pluiedev committed:
nix: bump nixpkgs version, fix build

Should switch to 25.11 once that releases.

We have to use nixpkgs-unstable for now since nixos-unstable hasn't
cached LLVM 20 for some reason (see NixOS/nixpkgs#440352)

pluiedev committed:
ci: fix Windows, Debian, Snap builds

Zig upstream changed how artifacts are called...

pluiedev committed:
ci: disable FreeBSD builds for now

reeBSD 14.3 does not yet ship Zig 0.15

pluiedev committed:
ci/build-nix: fix +version check

Zig 0.15's {t} formatter omits the type and only uses the tag name

pluiedev committed:
build: update build.zig.zon.{nix,txt,json}

pluiedev committed:
macos: fix build

pluiedev committed:
cli: add back libvaxis, all CLI commands ported

pluiedev committed:
datastruct: add our own copy of SegmentedList

SegmentedList was nuked from the Zig 0.15 standard library, but we still
need some pointer-safe data structure lying around :(

qwerasd205 commented:
Just adding an `if` before computing (at the location of the error there), like this:
```zig
if (entry.scale_factor == .adjustment)
    entry.scale_factor = .{
        .scale = self.scaleFactor(
            face.getMetrics(),
            entry.scale_factor.adjustment,
        ),
    };
```
should work

qwerasd205 commented:
Looking closer at that trace, actually, is it trying to scale the primary face against itself maybe?

qwerasd205 commented:
Ok, really funny behavior here-- when you have
```zig
some_union = .{
    .tag = fn_call(),
};
```
the tag is mutated on some_union before the fn_call is invoked. That's what's happening here. This behavior changed between 0.14.1 and 0.15.x. Here's a minimal repro:
```zig
const std = @import("std");

const T = union(enum) {
    a: u32,
    b: [4]u8,
};

fn doit(x: T) [4]u8 {
    std.debug.print("{}\n", .{x});
    return "erm?".*;
}

pub fn main() void {
    var x: T = .{
        .a = @bitCast([4]u8{'w', 'h', 'a', 't'}),
    };
    std.debug.print("{}\n", .{x});
    x = .{
        .b = doit(x),
    };
}
```

pluiedev commented:
what the hell

qwerasd205 commented:
So the solution is that we need to calculate the scale factor first and then set it on the entry, smth like
```zig
if (entry.scale_factor == .adjustment) {
    const scale =  self.scaleFactor(
        face.getMetrics(),
        entry.scale_factor.adjustment,
    );
    entry.scale_factor = .{
        .scale = scale,
    };
}
```

pluiedev committed:
font: calculate scale factor before updating entry

See https://github.com/ghostty-org/ghostty/pull/8372#issuecomment-3329702072

pluiedev committed:
build: explicitly refer to buildscript root in SharedDeps

pluiedev committed:
ci: disable Sentry in test-macos

pluiedev commented:
<img width="958" height="469" alt="image" src="https://github.com/user-attachments/assets/2012a964-8fe1-4bc5-afc0-109b375c0910" />

CI green :tada:

charlesrocket reviewed (commented):
Not seeing anything that would break on FBSD from the first read, though I haven't tested new deps. For FBSD CI, we could switch to binary pulls for now. I was thinking of pushing something like

```sh
ZIG_VERSION_LCL=0.15.1
cd /tmp
wget https://ziglang.org/download/$ZIG_VERSION_LCL/zig-x86_64-freebsd-$ZIG_VERSION_LCL.tar.xz
tar -xf zig-x86_64-freebsd-$ZIG_VERSION_LCL.tar.xz
alias zig=/tmp/zig-x86_64-freebsd-$ZIG_VERSION_LCL/zig
```
```git
+  #lang/zig \
